Add staking arithmetic dust doc (#66)
subspace/protocol-specs: docs/decex/staking_arithmetic_dust.md

There are 3 invariants in the protocol that should always hold:

## INVARIANT_1: `current_total_stake >= current_total_shares`

Initially, when the operator has just registered, and the operator owner is the first and only nominator, `current_total_stake:current_total_shares = 1:1`, and is equal to the operator's deposit. When the operator receives a reward, the reward is added to `current_total_stake`, while the `current_total_shares` remains unchanged. This increases the share price, so that `current_total_stake > current_total_shares`.

If the share price decreases during deposit/withdraw/unlock/slash, it means the reward (if any) is burned or stolen silently. And if the invariant is broken i.e. `current_total_stake < current_total_shares`, it means a nominator is losing its stake, and part of any new deposit will be burned immediately when joining the operator pool.

NOTE: Even during slashing, the operator's stake is fully slashed (not partially), and each nominator's stake and share will be removed from the pool, so slashing won't cause the share price to decrease.

## INVARIANT_2: `current_total_shares >= sum(nominator.shares)`

> The `current_total_shares` is expected to be the sum of all the nominator's shares in the pool (excluding pending deposits and withdrawals in the current epoch). But in practical, there is arithmetic dust, so the best we can do is to ensure the dust always flavors the protocol i.e. `current_total_shares >= sum(nominator.shares)`.

Reviewer: ```suggestion
`current_total_shares` is expected to be the sum of all the nominator's shares in the pool (excluding pending deposits and withdrawals in the current epoch). But in practice, there is arithmetic dust, so the best we can do is to ensure the dust always favours the protocol i.e. `current_total_shares >= sum(nominator.shares)`.
```

For the `INVARIANT_2` to hold:
```rust
stake_to_share(deposits_in_epoch) - D >= sum(stake_to_share(stake_i)) - D_i)

stake_to_share(deposits_in_epoch) - D >= sum(stake_to_share(stake_i)) - sum(D_i)

-D >= -sum(D_i)

D <= sum(D_i)
```

So the sum of the dust of all the individual nominator's stake-to-share conversion must be larger or equal to the dust of the operator's stake-to-share conversion, otherwise, `INVARIANT_2` will be broken. For `D <= sum(D_i)`, the share `sum(D_i) - D` is left in the pool that no one is entitled to withdraw and is given to the treasury after all nominators are unlocked.

### Withdraw: share-to-stake conversion

Within a given epoch, assuming there are `n` nominators withdraw `share_0, .., share_n` respectively:

```rust
// For individual nominator:
for share_i in  share_0, .., share_n {
    nominator.shares -= share_i;
    nominator.balance += share_to_stake(share_i) - D_i
}

// For the operator pool:
withdrawals_in_epoch = sum(share_0, .., share_n)

operator.current_total_shares -= withdrawals_in_epoch
operator.current_total_stake -= share_to_stake(withdrawals_in_epoch) - D
```

> The update of the withdrawn share on both the individual nominator and the operator pool are pure subtraction with no arithmetic dust. While the update of the stake is based on the share-to-stake conversion, and the only connection between the nominator unlocked stake and the `operator.current_total_stake` is the same share price that they use to convert withdrawn shares.

Reviewer: ```suggestion
Updating the withdrawn shares on both the individual nominator and the operator pool are pure subtraction, with no arithmetic dust. But updating the stake is based on the share-to-stake conversion, so the only connection between the nominator unlocked stake and the `operator.current_total_stake` is that they use the same share price to convert withdrawn shares.
```

> For the `INVARIANT_3` to hold:

Reviewer: ```suggestion
For `INVARIANT_3` to hold:
```

```rust
share_to_stake(withdrawals_in_epoch) - D >= sum(share_to_stake(share_i)) - D_i)

share_to_stake(withdrawals_in_epoch) - D >= sum(share_to_stake(share_i)) - sum(D_i)

-D >= -sum(D_i)

D <= sum(D_i)
```

> So the sum of the dust of all the individual nominator's share-to-stake conversion must be larger or equal to the dust of the operator's share-to-stake conversion, otherwise, `INVARIANT_3` will be broken. For `D <= sum(D_i)`, the stake `sum(D_i) - D` is burn.

Reviewer: ```suggestion
So the sum of the dust of all the individual nominator's share-to-stake conversion must be larger or equal to the dust of the operator's share-to-stake conversion, otherwise, `INVARIANT_3` will be broken. For `D <= sum(D_i)`, the stake `sum(D_i) - D` is burnt.
```

> NOTE: The nominator unlock and slash can be seen as a variant of full withdrawal where `n = 1` and `withdrawals_in_epoch = share_1 = nominator.shares` so `D = D_1`, `INVARIANT_3` is always held in these cases.

Reviewer: ```suggestion
NOTE: The nominator unlock and slash can be seen as a variant of full withdrawal where `n = 1` and `withdrawals_in_epoch = share_1 = nominator.shares` so `D = D_1`, `INVARIANT_3` always holds in these cases.
```
